example/simple_testing/src/lib.rs

```diff
@@ -0,0 +1,137 @@
+    env: JNIEnv<'local>,
+    callback_wrapper: JObject,
+) -> *mut DiplomatCallback {
+    // define the callback runner
+    extern "C" fn run_callback(data: *const c_void, args: *const c_void, runtime: *mut c_void) {
+        unsafe {
+            let runtime = runtime.cast::<ForeignRuntime>();
+            let data = data.cast::<GlobalRef>();
+
+            // attach_current_thread_permanently means that it attaches to the
+            // current JVM thread, and that if the JVM dies then this call also dies.
+            let jvm = match *runtime {
+                ForeignRuntime::JVM(ref jvm) => jvm,
+                _ => panic!("JVM callbacks need a JVM to run in"),
+            };
+            let mut env = jvm.attach_current_thread_permanently().unwrap();
+            // set the args pointer
+            env.call_method((*data).as_obj(), "set_args_pointer", "(J)V", &[jni::objects::JValueGen::Long(args as jlong)])
+                .unwrap();
+            // note: "run" is a method from Runnable, so we can't give it an arg
+            env.call_method((*data).as_obj(), "run", "()V", &[])
+                .unwrap();
+        }
+    }
+    // define the destructor for the callback wrapper
+    extern "C" fn destructor(this: *const c_void) -> bool {
+        unsafe {
+            // TODO free the memory
+        }
+        true
+    }
+
+    // initialize the callback wrapper
+    let jvm = env.get_java_vm().unwrap();
+    let ref_to_cb_wrapper = env.new_global_ref(callback_wrapper).unwrap();
+    let ret = DiplomatCallback {
+        data: Box::into_raw(Box::new(ref_to_cb_wrapper)) as _,
+        run_callback,
+        destructor,
+        foreign_runtime: Box::into_raw(Box::new(ForeignRuntime::JVM(jvm))) as _,
+    };
+
+    Box::into_raw(Box::new(ret))
+}
+
+// ------------------------------------------------- this next section is specific to the callback, and goes in the trampoline crate
+
+// wrapper struct for the args of this specific function
+// note: this will need to be opaque so it can be passed by ref to the foreign language
+pub struct FFIArgsWrapper_f {
+    arg0: i32
+}
+
+impl FFIArgsWrapper_f {
+    pub fn get_arg0(&self) -> i32 {
+        self.arg0
+    }    
+}
+
+impl FFIArgsWrapper for FFIArgsWrapper_f {
+    fn num_args(&self) -> usize {
+        1
+    }
+}
+
+// function to get the first arg out of the arg struct for this specific callback
+#[no_mangle]
+pub unsafe extern "C" fn get_arg0_from_ffiargswrapper_f_pointer(fw: *const FFIArgsWrapper_f) -> i32 {
+    unsafe {
+        (*fw).get_arg0()
+    }
+}
+
+// function that actually calls the callback
+// it's basically a heavily instrumented version of the original Rust function that takes
+// a callback as an argument.
+// it wraps the arguments to the callback in the arg wrapper struct, and calls the callback
+#[no_mangle]
+pub unsafe extern "C" fn GEND_BRIDGE_test_run_fn(cb: &DiplomatCallback) {
```

Comment:
in Diplomat, this function would be called `nameoftype_nameofmethod`. The name initially confused me, I thought this was the trampoline.